Repo URL: https://github.com/<redacted>/assignment-4-equation-of-a-slime-<redacted>

Commit ID: f05499f

In [1]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn import linear_model
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import ShuffleSplit
from sklearn.model_selection import KFold
from sklearn.preprocessing import PolynomialFeatures

## Part 1. Loading the dataset

In [2]:
# Using pandas load the dataset (load remotely, not locally)
dataset = pd.read_csv("science_data_large.csv")

# Output the first 15 rows of the data
print(dataset.head(15))

# Display a summary of the table information (number of datapoints, etc.)
print(dataset.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## Part 2. Splitting the dataset

In [3]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = dataset.iloc[:,:-1]
y = dataset.iloc[:,-1]

print(X)
print(y)
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.1, random_state=42)

     Temperature °C  Mols KCL
0               469       647
1               403       694
2               302       975
3               779       916
4               901        18
..              ...       ...
995             894       847
996             327       982
997             791       213
998             769       553
999             919       452

[1000 rows x 2 columns]
0      6.244743e+05
1      5.779610e+05
2      6.196847e+05
3      1.460449e+06
4      4.325726e+04
           ...     
995    1.545661e+06
996    6.737041e+05
997    3.477543e+05
998    8.684794e+05
999    8.476413e+05
Name: Size nm^3, Length: 1000, dtype: float64


## Part 3. Perform a Linear Regression

In [4]:
# Use sklearn to train a model on the training set
model = LinearRegression()
train_model = model.fit(X_train, y_train)

# Create a sample datapoint and predict the output of that sample with the trained model
sample_data = pd.DataFrame({'Temperature °C': [500], 'Mols KCL': [333]})
predicted_output = model.predict(sample_data)

print(f"Predicted output for the sample point {sample_data.values[0]}: {predicted_output[0]:.2f}")

# Report on the score for that model, in your own words (markdown, not code) explain what the score means
r2_score = model.score(X_test, y_test)
print("r2 score:", r2_score)

# Extract the coefficents and intercept from the model and write an equation for your h(x) using LaTeX
coefficients = model.coef_
intercept = model.intercept_

print("Coefficients:", coefficients)
print("Intercept:", intercept)

Predicted output for the sample point [500 333]: 367569.18
r2 score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764




$h(x) = -409391.47958340764 + 866.14641337(x_1) + 1032.69506649(x_2)$

# Report on the score for that model, in your own words (markdown, not code) explain what the score means

Answer: The $r^2$ score ended up being around 0.8552. The $r^2$ score is representative of how well the model predicts labels in the test data (in accordance to its features). More formally, it is calculated using the formula (1 - Sum Sqaured Residuals). In class, we learned that a residual is how far from a point the point's predicted value is. '1.0' is the maximum $r^2$ you can recieve, it is the value we are aiming for. Our score for this linear regreession model is about 0.8552. This actually isn't a great score, as computers in today's world are capable of achieving great numbers. A score of 0.8552 indicates that the predictions are innacurate and there is still room for improvement. This could mean that we may need more datapoints, or maybe a linear regression isn't fit for this dataset.

## Part 4. Use Cross Validation

In [5]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
n_splits = 5  # Number of folds
kf = KFold(n_splits=n_splits, shuffle=True)

# Perform cross-validation and calculate R² scores
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='r2')

print(f"Shuffle 1: Cross-Validation R^2 Scores: {cv_scores}")
print(f"Shuffle 1: Mean R^2 Score: {np.mean(cv_scores):.4f}")
print(f"Shuffle 1: Standard Deviation of R^2 Scores: {np.std(cv_scores):.4f}")
print("")

n_splits = 5 
kf = KFold(n_splits=n_splits, shuffle=True)

# Perform cross-validation and calculate R² scores
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='r2')

print(f"Shuffle 2: Cross-Validation R^2 Scores: {cv_scores}")
print(f"Shuffle 2: Mean R^2 Score: {np.mean(cv_scores):.4f}")
print(f"Shuffle 2: Standard Deviation of R^2 Scores: {np.std(cv_scores):.4f}")
print("")

n_splits = 10  
kf = KFold(n_splits=n_splits, shuffle=True)

# Perform cross-validation and calculate R² scores
cv_scores = cross_val_score(model, X, y, cv=kf, scoring='r2')

print(f"Shuffle 3: Cross-Validation R^2 Scores: {cv_scores}")
print(f"Shuffle 3: Mean R^2 Score: {np.mean(cv_scores):.4f}")
print(f"Shuffle 3: Standard Deviation of R^2 Scores: {np.std(cv_scores):.4f}")

# Report on their finding and their significance



Shuffle 1: Cross-Validation R^2 Scores: [0.87301845 0.85762739 0.83937317 0.85857513 0.86314277]
Shuffle 1: Mean R^2 Score: 0.8583
Shuffle 1: Standard Deviation of R^2 Scores: 0.0109

Shuffle 2: Cross-Validation R^2 Scores: [0.81888104 0.8674093  0.87317482 0.87479761 0.85556193]
Shuffle 2: Mean R^2 Score: 0.8580
Shuffle 2: Standard Deviation of R^2 Scores: 0.0207

Shuffle 3: Cross-Validation R^2 Scores: [0.86597892 0.87773936 0.85081553 0.88874943 0.8513354  0.88785041
 0.85273005 0.83767331 0.86506776 0.80912633]
Shuffle 3: Mean R^2 Score: 0.8587
Shuffle 3: Standard Deviation of R^2 Scores: 0.0229


# Report on their finding and their significance

Answer: I ran three shuffles of the data with split values of 5 and 10. And the mean of the $r^2$ scores across all the subsets of data, across all the shuffles, turned out to be quite similar. This is a good thing, because it tells us that our prediction model is consistent. Moreover, the standard deviation across each shuffle is quite low, which is also great because it also emphasizes the consistency of our model. Consistency is important because we want our model to perform well when it encounters new data, and consistency is key in identifying how much you can trust your model.

## Part 5. Using Polynomial Regression

In [6]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
poly = PolynomialFeatures(degree=2)
X_train_poly = poly.fit_transform(X_train)

model_poly = LinearRegression()
train_model_poly = model_poly.fit(X_train_poly, y_train)

X_test_poly = poly.transform(X_test)

y_pred = model_poly.predict(X_test_poly)

r2 = model_poly.score(X_test_poly, y_pred)

print(f"R² Score: {r2:.2f}")

# Report on the metrics and output the resultant equation as you did in Part 3.
coefficients_poly = model_poly.coef_
intercept_poly = model_poly.intercept_
print("")
print("Coefficients:", coefficients_poly)
print("Intercept:", intercept_poly)

R² Score: 1.00

Coefficients: [ 0.00000000e+00  1.20000000e+01 -1.27195488e-07  1.26494371e-11
  2.00000000e+00  2.85714287e-02]
Intercept: 2.0477105863392353e-05


=> $p(x) = 2.0477105863392353^{-05} + 0.00000(x_1) + 1.200000^{01}(x_2) -1.27195488^{-07}(x_3) + 1.26494371^{-11}(x_4) + 2.00000000^{00}(x_5) + 2.85714287^{-02}(x_6)$

=> $p(x) = 2.0477105863392353^{-05} + 1.200000^{01}(x_2) + 2.00000000(x_5) + 2.85714287^{-02}(x_6)$

=> $p(x) = 0.00002 + 12(x_2) + 2(x_5) + 0.02857(x_6)$

# Report on the metrics and output the resultant equation as you did in Part 3.

Answer: I was very surprised to find that our $r^2$ score ended up being 1.0 after conducting a polynomial regression. This is very impressive and sounds too good to be true. I even had to double check to make sure I had not run the predictions on the training data itself, and it seems like I didn't. This tells us that a polynomial regression is perfect for this dataset. A score of '1.0' tells us that the model makes perfect predictions for each data point. Therefore, our endeavor was a success and we can now use this model/equation to inhabitate Mars.